src/field_screen_effect.c

```diff
@@ -73,3 +86,257 @@ void SetFlashScanlineEffectWindowBoundaries(u16 *dest, s32 centerX, s32 centerY,
+    gTasks[taskId].data[tDirection] = DIR_WIPE_IN;
+}
+
+void DoOutwardBarnDoorWipe(void)
+{
+    u8 taskId = CreateTask(Task_BarnDoorWipe, 80);
+    gTasks[taskId].data[tDirection] = DIR_WIPE_OUT;
+}
+
+static void BarnDoorWipeSaveGpuRegs(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+    data[0] = GetGpuReg(REG_OFFSET_DISPCNT);
+    data[1] = GetGpuReg(REG_OFFSET_WININ);
+    data[2] = GetGpuReg(REG_OFFSET_WINOUT);
+    data[3] = GetGpuReg(REG_OFFSET_BLDCNT);
+    data[4] = GetGpuReg(REG_OFFSET_BLDALPHA);
+    data[5] = GetGpuReg(REG_OFFSET_WIN0H);
+    data[6] = GetGpuReg(REG_OFFSET_WIN0V);
+    data[7] = GetGpuReg(REG_OFFSET_WIN1H);
+    data[8] = GetGpuReg(REG_OFFSET_WIN1V);
+}
+
+static void BarnDoorWipeLoadGpuRegs(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+    SetGpuReg(REG_OFFSET_DISPCNT, data[0]);
+    SetGpuReg(REG_OFFSET_WININ, data[1]);
+    SetGpuReg(REG_OFFSET_WINOUT, data[2]);
+    SetGpuReg(REG_OFFSET_BLDCNT, data[3]);
+    SetGpuReg(REG_OFFSET_BLDALPHA, data[4]);
+    SetGpuReg(REG_OFFSET_WIN0H, data[5]);
+    SetGpuReg(REG_OFFSET_WIN0V, data[6]);
+    SetGpuReg(REG_OFFSET_WIN1H, data[7]);
+    SetGpuReg(REG_OFFSET_WIN1V, data[8]);
+}
+
+void Task_BarnDoorWipe(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+    switch (data[9])
+    {
+        case 0:
+            BarnDoorWipeSaveGpuRegs(taskId);
+            SetGpuRegBits(REG_OFFSET_DISPCNT, DISPCNT_WIN0_ON);
+            SetGpuRegBits(REG_OFFSET_DISPCNT, DISPCNT_WIN1_ON);
+            if (data[10] == 0)
+            {
+                SetGpuReg(REG_OFFSET_WIN0H, 0);
+                SetGpuReg(REG_OFFSET_WIN1H, 0xF0FF); // 255.
+                SetGpuReg(REG_OFFSET_WIN0V, 255);
+                SetGpuReg(REG_OFFSET_WIN1V, 255);
+            }
+            else
+            {
+                SetGpuReg(REG_OFFSET_WIN0H, 120);
+                SetGpuReg(REG_OFFSET_WIN0V, 255);
+                SetGpuReg(REG_OFFSET_WIN1H, 0x78FF); // 255.
+                SetGpuReg(REG_OFFSET_WIN1V, 255);
+            }
+            SetGpuReg(REG_OFFSET_WININ, 0);
+            SetGpuReg(REG_OFFSET_WINOUT, WINOUT_WIN01_BG_ALL | WINOUT_WIN01_OBJ | WINOUT_WIN01_CLR);
+            data[9] = 1;
+            break;
+        case 1:
+            CreateTask(Task_BarnDoorWipeChild, 80);
+            data[9] = 2;
+            break;
+        case 2:
+            if (FuncIsActiveTask(Task_BarnDoorWipeChild) == 0)
+            {
+                data[9] = 3;
+            }
+            break;
+        case 3:
+            BarnDoorWipeLoadGpuRegs(taskId);
+            DestroyTask(taskId);
+            break;
+    }
+}
+
+// TODO: Find an implicit way to generate lhs/rhs shifts.
+static void Task_BarnDoorWipeChild(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+	u8 parentTaskId = FindTaskIdByFunc(Task_BarnDoorWipe);
+    u32 lhs, rhs;
+    if (gTasks[parentTaskId].data[tDirection] == DIR_WIPE_IN)
+    {
+        lhs = (u16)data[tChildOffset];
+        rhs = (u16)(240 - lhs);
```

Comment:
maybe these are s16? The compiler sometimes do unsigned operations over signed variables for optimization

Reply:
I think it's very likely. I've spent enough time on this function, so I'm not going to look into it further, but it doesn't work if you simply change to `s16` because there are some places where I actually rely on shifting into those unused upper 16 bits. And, in particular, this code relies on both the right shifts occurring after the left shifts:
```
lhs  = lhs2 >> 16;
rhs = rhs2 >> 16;
```

Reply:
```C
static void Task_BarnDoorWipeChild(u8 taskId)
{
    s16 *data = gTasks[taskId].data;
	u8 parentTaskId = FindTaskIdByFunc(Task_BarnDoorWipe);
    s16 lhs, rhs;
    if (gTasks[parentTaskId].tDirection == DIR_WIPE_IN)
    {
        lhs = tChildOffset;
        rhs = 240 - tChildOffset;
        if (lhs > 120)
        {
            DestroyTask(taskId);
            return;
        }
    }
    else
    {
        lhs = (120 - tChildOffset);
        rhs = (120 + tChildOffset);
        if (lhs < 0)
        {
            DestroyTask(taskId);
            return;
        }
    }
    SetGpuReg(REG_OFFSET_WIN0H, lhs);
    SetGpuReg(REG_OFFSET_WIN1H, WIN_RANGE(rhs, 240));
    if (lhs <= 89)
    {
        tChildOffset += 4;
    }
    else
    {
        tChildOffset += 2;
    }
}
```
check this one

Reply:
Brilliant! Thank you! I put your name on the commit. (I think I checked the box that allows you to commit to my branch, so feel free to do it yourself in the future)